trlx/model/accelerate_ppo_model.py

```diff
@@ -119,100 +141,42 @@ def loss(
             1.0 + self.config.method.cliprange,
         )
 
-        pg_loss = torch.mean(torch.max(pg_losses, pg_losses2))
+        pg_loss = torch.sum(torch.max(pg_losses, pg_losses2) * mask) / mask.sum()
+        loss = pg_loss + self.config.method.vf_coef * vf_loss
+
+        stats = {
+            "loss": loss,
+            "pg_loss": pg_loss,
+            "vf_loss": vf_loss,
+        }
 
-        model_loss = pg_loss + self.config.method.vf_coef * vf_loss
-        return model_loss, pg_loss, vf_loss
+        return loss, stats
 
     def post_epoch_callback(self):
-        # TODO(dahoas): are experiences being made for dataloaders on each process or same dataloader
-        self.epoch += 1
         self.store.clear_history()
         self.orch.make_experience(
             self.config.method.num_rollouts, self.iter_count
         )  # Collect more rollouts for training
 
     def post_backward_callback(self):
-        batch = self.logs["batch"]
         # Update kl_coefficient
         self.kl_ctl.update(self.mean_kl, self.config.train.batch_size)
-        # Run evaluation
-        if self.accelerator.is_main_process:
-            if (
-                self.iter_count % self.config.train.eval_interval == 0
-                or self.iter_count <= self.config.method.ppo_epochs
-            ):
-                text = self.tokenizer.batch_decode(batch.query_tensors)
-                eval_batch: PromptBatch = PromptBatch(
-                    text=text, tokens=batch.query_tensors
-                )
-                query_tensors, response_tensors, response_text = self.act(eval_batch)
-                gen_texts = [q + r for q, r in zip(eval_batch.text, response_text)]
-                scores = self.orch.score(gen_texts)
-                mean_score = torch.mean(scores).item()
-                rows = list(zip(gen_texts, scores.tolist()))
-                stats = {
-                    "mean_score": mean_score,
-                    "responses": wandb.Table(columns=["response", "score"], rows=rows),
-                    "pg_loss": self.logs["pg_loss"],
-                    "vf_loss": self.logs["vf_loss"],
-                    "kl_coef": self.kl_ctl.value,
-                }
-                self.accelerator.log(stats, step=self.iter_count)
-                self.accelerator.print(
-                    "Step: {}, Mean score: {}, pg_loss: {}, vf_loss: {}, kl_coef: {}".format(
-                        self.iter_count,
-                        mean_score,
-                        stats["pg_loss"],
-                        stats["vf_loss"],
-                        self.kl_ctl.value,
-                    )
-                )
-
-    def learn(self, log_fn=None, save_fn=None, eval_fn=None):
```

Comment:
Does the PPO implementation still have multiple ppo_epochs per batch? I see a new variable defining this (n_updated_per_batch) but since we are relying on the base_model's training loop I am not seeing where it gets used.

Perhaps if we do not want to override base_model's learn method we should write something in the post_backward callback